src/lib/crypto/snarky_tests/snarky_tests.ml

```diff
@@ -0,0 +1,274 @@
+(* this is ugly, but [Core_kernel.String] doesn't have a [trim] function *)
+let trim = String.trim
+
+open Core_kernel
+
+(* instantiate snarky for Vesta *)
+
+module Impl = Snarky_backendless.Snark.Run.Make (struct
+  include Kimchi_backend.Pasta.Vesta_based_plonk
+  module Inner_curve = Kimchi_backend.Pasta.Pasta.Pallas
+end)
+
+(* helpers *)
+
+let compare_with obtained filepath =
+  let (_ : int) = Sys.command "tree" in
+  let filepath = "examples/" ^ filepath in
+  let expected = In_channel.read_all filepath in
+  if String.(trim obtained <> trim expected) then (
+    Format.printf "mismatch for %s detected\n" filepath ;
+    Format.printf "expected:\n%s\n\n" expected ;
+    Format.printf "obtained:\n%s\n" obtained ;
+    failwith "circuit compilation has changed" )
+
+let check_json ~input_typ ~return_typ ~circuit filename () =
+  let cs : Impl.R1CS_constraint_system.t =
+    Impl.constraint_system ~input_typ ~return_typ circuit
+  in
+  let serialized_json =
+    Kimchi_backend.Pasta.Vesta_based_plonk.R1CS_constraint_system.to_json cs
+  in
+  compare_with serialized_json filename
+
+(* monadic API tests *)
+
+(** Both the monadic and imperative API will produce the same circuit hash. *)
+let expected = "5357346d161dcccaa547c7999b8148db"
+
+module MonadicAPI = struct
+  module Impl = Snarky_backendless.Snark.Make (struct
+    include Kimchi_backend.Pasta.Vesta_based_plonk
+    module Inner_curve = Kimchi_backend.Pasta.Pasta.Pallas
+  end)
+
+  let main ((b1, b2) : Impl.Boolean.var * Impl.Boolean.var) =
+    let open Impl in
+    let%bind x = exists Boolean.typ ~compute:(As_prover.return true) in
+    let%bind y = exists Boolean.typ ~compute:(As_prover.return true) in
+    let%bind z = Boolean.(x &&& y) in
+    let%bind b3 = Boolean.(b1 && b2) in
+
+    let%bind () = Boolean.Assert.is_true z in
+    let%bind () = Boolean.Assert.is_true b3 in
+
+    Checked.return ()
+
+  let get_hash_of_circuit () =
+    let input_typ = Impl.Typ.tuple2 Impl.Boolean.typ Impl.Boolean.typ in
+    let return_typ = Impl.Typ.unit in
+    let cs : Impl.R1CS_constraint_system.t =
+      Impl.constraint_system ~input_typ ~return_typ main
+    in
+    let digest = Md5.to_hex (Impl.R1CS_constraint_system.digest cs) in
+    Format.printf "expected:\n%s\n\n" expected ;
+    Format.printf "obtained:\n%s\n" digest ;
+    assert (String.(digest = expected))
+end
+
+(* circuit-focused tests *)
+
+module BooleanCircuit = struct
+  module Request = struct
+    type _ Snarky_backendless.Request.t +=
```

Comment:
It would be good to split these further into e.g. a handler test, an exists test, etc. As it stands, it's rather difficult to track what is and isn't tested.